components/pkg/utils/snowflake/ringbuffer.go

```diff
@@ -0,0 +1,186 @@
+//
+// Copyright 2021 Layotto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package snowflake
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"mosn.io/pkg/log"
+)
+
+const (
+	CAN_TAKE_FLAG = 1
+	CAN_PUT_FLAG  = 0
+)
+
+//avoid false sharing
+type PaddedInt struct {
+	_     [7]int64
```

Comment:
explain this

Reply:
cache line padding. You can refer to this: https://github.com/baidu/uid-generator/blob/master/README.zh_cn.md